ssc/Pos/Ssc/Worker.hs

```diff
@@ -365,14 +380,19 @@ waitUntilSend msgTag epoch slMultiplier = do
             -- TODO [CSL-2173]: Clarify
             leftToPanic "waitUntilSend: " $
             mkLocalSlotIndex $ slMultiplier * fromIntegral slotSecurityParam
-    Timestamp beginning <-
-        getSlotStartEmpatically $
-        SlotId {siEpoch = epoch, siSlot = slot}
-    curTime <- currentTime
-    let minToSend = curTime
-    let maxToSend = beginning + mpcSendInterval
-    when (minToSend < maxToSend) $ do
-        let delta = maxToSend - minToSend
+    timeDiff <- do
+        -- Compute an approximation for the time difference between the current
+        -- slot and the next slot. The result might be inaccurate, i tak soydet.
+        curSlot <- getCurrentSlotInaccurate
+        slotDur <- getNextEpochSlotDuration
+        let
+            targetSlot = SlotId {siEpoch = epoch, siSlot = slot}
+            subzero a b = if a > b then a - b else 0
+            slotDiff = (subzero `on` flattenSlotId) targetSlot curSlot
+        return $ fromMicroseconds $
+            fromIntegral slotDiff * toMicroseconds slotDur
+    let delta = timeDiff + mpcSendInterval
+    when (delta > 0) $ do
```

Comment:
Now I see that semantics of this code changed in a bad way. `delta` will always be `mpcSendInterval`, because `waitUntilSend` is always called after `targetSlot` (which is probably not the best name) starts.
`targetSlot` is the slot when some phase starts (e. g. opening phase starts at `4k`-th slot). When this phase starts, we'll call `onNewSlotOpening` which will call `waitUntilSend`.
So in fact `slotDiff` should be `curSlot - targetSlot`, `delta` should be `mpcSendInterval - timeDiff` and some variables should be renamed.
`targetSlot` can be called `phaseStartSlot`, `timeDiff` can still be called `timeDiff`, but it should be commented (that it's the different between beginning of the current slot and `phaseStartSlot`).